### DISCUSSION TITLE
Refresh data from the form

### DISCUSSION THREAD

**Discussion by thurfir** [react-hook-form/react-hook-form#2149] [Q&A] [answered]
Hi, is there a method to refresh the react-hook-form data from its HTML?

I'm trying to apply a style to an input's parent element when it is not empty with useWatch; it mostly works except when the form is preloaded by the browser (login/passwords, input [autocomplete attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete)).

**Comment by bluebill1049** [comment]
what do you mean? do you have a CSB?

**Comment by thurfir** [comment]
Sorry, what does the term 'CSB' means ?

I'm talking about browsers that often pre-fill a form for its users:

- if the form contains a type="password" input, the browser pre-fills the login + password inputs.

- if the form contains input with an "autocomplete" attribute, the browser pre-fill it with the user adress, phone, email, country, ... according to the attribute value.

- if I go back to a page (previous button), the browser retrieves the form content from its local cache.

Currently, these pre-fills are not reflected in the react-hook-form's data until a change is made (useWatch return the default value). I'd like to trigger a refresh to update that data and update my react content accordingly (in my above example, add a specific css class).

**Reply by bluebill1049** [reply]
CSB: codesandbox. wondering if this is browser level stuff or the lib?

**Comment by thurfir** [comment]
Hi @bluebill1049 , here is a sandbox. I used the login + password : probably the easiest way to test it as autocomplete depends of the browser's saved data.

https://codesandbox.io/s/react-hook-form-test-login-password-25jpu?file=/src/index.js

In Chome (windows), open the sandbox with the new window mode (inputs are never pre-filled inside the embed). Type any login + password and submit, and save if prompted. Refresh (F5) the page, the browser shall have pre-filled your inputs:

![image](https://user-images.githubusercontent.com/22148896/87281110-2ed48c80-c4f3-11ea-9b3b-250d4b1264f7.png)

Click anywhere, the _watch_ catches up with the form.

![image](https://user-images.githubusercontent.com/22148896/87281270-575c8680-c4f3-11ea-833c-3f6d13b6e3f0.png)

**Reply by bluebill1049** [reply]
This is browser level stuff @thurfir, don't think RHF can do much about it.

**Reply by thurfir** [reply]
It sure is. It seems there are some way to detect an autofill, but it seems more like R&D at this point: https://paul.kinlan.me/detecting-when-autofill-happens/

I am mostly looking at mitigating this issue (it impact my page styles) by forcing RHF to refresh its data from the page, is there a way to do that?

**Reply by bluebill1049** [reply]
what do you mean `forcing RHF to refresh its data from the page`?

**Reply by thurfir** [reply]
Just like what the click in my above demo does: to force a re-render with the _watch_ value updated.

**Reply by bluebill1049** [reply]
setValue at useEffect?

**Reply by thurfir** [reply]
Does setValue works without specifying a value? I'd like to read the input, not override it.

**Reply by bluebill1049** [reply]
hmmm good point.

**Reply by thurfir** [reply]
In essence I'm looking for a "reverse-reset": the _reset_ method apply the data to the HTML, I'd like to apply the HTML to the data ;-)

You probably already have that, at least internally, for _handleSubmit_ and the like.

**Reply by bluebill1049** [reply]
that sounds like `getValues()`.

**Comment by thurfir** [comment]
I you want I can open an Issue if there is no current "refresh" method ?

**Reply by bluebill1049** [reply]
sorry, no, we don't have plan to introduce `refresh` API.

**Comment by thurfir** [comment, marked as answer]
I tried getValues() as you suggested, I doesn't work: https://codesandbox.io/s/react-hook-form-test-login-password-spkww?file=/src/index.js (called here every 100ms)

![image](https://user-images.githubusercontent.com/22148896/87535360-a9d6a800-c697-11ea-99e6-4e40710a3db0.png)

What do you suggest?

**Reply by bluebill1049** [reply]
that's what browser stored. don't think its readable.

**Reply by thurfir** [reply]
Oh my, you are right, even with a simple querySelector the data cannot be reached!

 Nothing that you can do, but I guess we'll have to wait until CSS 4 to properly style inputs' parents when they contain some data.
Thanks for your help!

**Comment by Manubi** [comment]
Hi,
is there any update on this issue? As I'd like to use the html autoComplete tag. But as said above, if you use the autocomplete feature react-hook-form isn't able to get the values unless you click in the field. 

Thanks!

![CleanShot 2021-04-17 at 18 28 17@2x](https://user-images.githubusercontent.com/20935138/115119924-183ff200-9fab-11eb-952c-7ffdc0c27f14.png)
